### ISSUE TITLE
Should we load firmware directly from the device's vendor and modem partitions?

### ISSUE THREAD

**Issue by zhuowei** [postmarketOS/pmbootstrap#637]
Many Android devices made after the Nexus 9 (so post 2014) store most of their proprietary firmware on the /vendor partition, and, at least on the Nexus 6P and probably many other Qualcomm devices, the modem firmware is stored on a modem partition (usually mounted at /firmware).

For these devices, should we mount these partitions in pmOS and load blobs from there - which would give us a smaller system image and avoid any license issue with redistributing blobs? Or should we bundle the blobs directly in the system image, so that people don't have to flash a /vendor and a modem firmware from the correct stock firmware in order to start pmOS?

**Comment by drebrez** [comment]
@zhuowei good question, I have a samsung i9070 (from 2012) and also have a `/modem` partition (I suppose contains the blobs for the momem).
I wanted to start trying to enable the modem and like you I didn't know if simply mount that partition or get the blobs from somewhere else.
@ollieparanoid can probably help us answering this question.

**Comment by ollieparanoid** [comment]
I recommended in the past that we package the firmware (not sure if here in the tracker, or in the channel), for the following reasons:
* We know which version we have
* We can update it easily (either with official new versions, or with hacks like nexmon)
* We don't run into conflict with other OSes that may be installed on the device (e.g. when uninstalling pmOS and installing another ROM, or when doing the NFS boot like you do)

You are right with the licensing issues, we should look into that before releasing official images or official packages of the firmware files.

**Comment by PabloCastellano** [comment]
Now that I know how to make wifi work in titan, I have faced this same question.

In my case the firmware files are spread among different partitions and I mounted the firmware partitions in /mnt like:
* `/mnt/data`
* `/mnt/persist`
* `/mnt/firmware`

And then symlink the files from `/lib/firmware/postmarketos`.

I have a branch working with this approach. After that I found this open issue :smile:.

@ollieparanoid reasons to package firmware files make totally sense and therefore we only have to decide what to do with the licensing issue. Asking the manufacturer for clarification is an option but sincerely, based on my past experiences I don't have so much hope on it (either for having their answer or their permission to redistribute the blobs).

**Comment by ollieparanoid** [comment]
I agree with you, the "asking the manufacturer" path is unlikely to be working out, especially if we want to patch the firmware files to fix security holes and maybe enable monitoring mode (#592).

In my opinion, we don't have a problem here unless we provide binary packages (which we don't need to do in the current state of the project). And when postmarketOS gets close to being a daily driver, we could do it like such issues are handled on most distributions: Create a package, that does not include the firmware files directly, but somehow downloads and patches them from the internet after installation.

**Comment by PabloCastellano** [comment]
Looks like the discussion is ended. I would document in the wiki what we have agreed here.

How to load firmware blobs?
- First, search some repo which already have them
- If you cannot find it, host them yourself
- Explain [advantages](https://github.com/postmarketOS/pmbootstrap/pull/771#issuecomment-336643976) of downloading vs using the ones in your vendor partition

**Comment by ollieparanoid** [comment]
I have added an entry to the FAQ, which directly answers the question:
https://wiki.postmarketos.org/wiki/FAQ#Why_don.27t_we_use_firmware_files_from_Android.27s_firmware_partition.3F

Feel free to add more documentation, for now it's recommended to look at the packaging of other firmware files.

**Comment by mirh** [comment]
Ehrm.. I don't want to be that usual killjoy, but at least on very low end devices you'd want to keep install size as low as possible (internal space comes at a premium, and sdcard might not be the best idea to use for system)

"Recycling" those few tens of MB could make quite a difference (and I'm not sure why patching still couldn't be done, if really needed - nor why this wouldn't be a godsend for all the distribution licenses paranoia)

**Comment by mirh** [comment]
No firmware partition means no firmware elsewhere than system image. 
I'm not sure what's even the deal there. 

Speaking instead of "touching" firmware partition.. It depends from phone to phone*, but I cannot see why (if we are imaging of *using that space better*) tinkering with the partition should be forbidden for updating then. 
If the only concern is compatibility with other/original OS.. I thought firmware was "nice" because it's "feed" directly to the hardware component, and "software on cpu" has nothing to say about it?

Moreover, nexmon-like-derived** tweaks aside.. I'm not sure whoever would have been producing firmware for a device, if not its original OEM?

*I mean, I dunno, but for example I wouldn't be surprised if h-boot/radio were somehow intertwined and protected on s-on htc devices
**which I'd be expecting to strive NOT to break a thing like ever

**Comment by ollieparanoid** [comment]
> I wouldn't be surprised if h-boot/radio were somehow intertwined and protected on s-on htc devices

That sounds like a deal breaker TBH.

**Comment by mirh** [comment]
Nexmon just does wifi last time I checked (which, by the way, don't have open licenses ofc, but are [pretty](Broadcom/Cypress) [duper](https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/tree/LICENCE.ti-connectivity) [fine](https://github.com/kvalo/ath10k-firmware/blob/master/LICENSE.qca_firmware) with redistribution)
AFAIK partitions, and what I thought this thread was about, are just for *modem*, as in GSM/UMTS. 

> As I see it, we have a good solution now, that works for all devices,

Of course, I can even understand "inertia" as a valid motive. 
But what I'm saying, is that there's no technical reason *not to* use partitions straight away as they are (just like android might already do, for example)
And if we look at the legal side of the thing (not that I personally care of, but still I know you quite do) there are possibly reasons *for* this practice, given I'm not aware of them ever being redistributed outside official firmwares

**Comment by ollieparanoid** [comment]
> Nexmon just does wifi last time I checked

Sorry for that, I've missed the "radio/baseband" part of your message.

> AFAIK partitions, and what I thought this thread was about, are just for modems, as in GSM/UMTS.

I did not realize that you wanted to differentiate between baseband modem firmware files and other firmware files. Isn't that more complicated than treating all firmware files the same way?

> But what I'm saying, is that there's no technical reason *not to* use partitions straight away as they are (just like android might already do, for example)

I thought I have listed _three_ technical advantages for not using the firmware partition in https://github.com/postmarketOS/pmbootstrap/issues/637#issuecomment-332599378. To summarize, it boils down to be in control of the files.

> Of course, I can even understand "inertia" as a valid motive.

You know how this project works: there's no boss that tells everyone what to do, instead we have a community that makes decisions together after discussing technical arguments like we are right now. So there would not be a problem with changing the decision that has been made about shipping the firmware files as packages vs. using them from existing partitions if everyone thinks that's the better way.

But in order to do that, we would need to make sure that the new solution is better than what we have now, and I don't see it. I'm not even sure if you are arguing for **not ever replacing the firmware blobs** (bad because we can't make security updates if we should get to a point in the future where the community is able to patch all firmware blobs, even for modems), or if we **should be able to replace them** (then your licensing argument does not make sense).

To follow through with this:
* What exactly are you proposing?
* Why is it better than what we have now?
* How would we implement this?

**Comment by mirh** [comment]
> I did not realize that you wanted to differentiate between baseband modem firmware files and other firmware files. Isn't that more complicated than treating all firmware files the same way?

Ok sorry, I just realized now that OP was talking about *newer phones* [mess](https://github.com/sonyxperiadev/device-sony-yoshino/blob/master/rootdir/vendor/etc/fstab.yoshino)
My mind was left to the old way of functioning, where the only code you could find in extra partitions was only* radio firmware *if any*. 

In this case, I never really looked at what's inside /vendor partition. So I cannot really comment on the state of affairs (let alone today, with [project treble](https://www.xda-developers.com/how-project-treble-revolutionizes-custom-roms-android-oreo/) moving even more dependencies in there)

> To summarize, it boils down to be in control of the files.

Yes, but my point was (is?) that you can be in control of stuff in other partitions too. 
With no supposed breakage whatsoever, at least in the bare theory, if that's really \*firmware\*, and not *blobs*. 

> or if we should be able to replace them (then your licensing argument does not make sense).

You *can* replace them, and you *can* update them**. 
(Now, putting aside again that wifi firmware licenses are discretely lax) Nexmon itself can be installed on the devices, correct?
If you don't ship the patched blobs (or even the blobs at all), and it's the user himself that "tinker for its own personal use" then I think you are as safe as you can possibly be. 

Then, perhaps you are being confused by my points all being "negative" - but this is just because I fear I may miss something and saying "never ship yourself fw if it's already there" (for as much as I can hardly think to a realistic situation where it would make sense) seems bullish. 
Cheers. 

\*I mean.. I could also think of Samsung's /preload partition then.. But I wouldn't even be sure of that being used by stock rom(*!*)
**and now I understand why you brought up nexmon, even though it didn't seem relevant

**Comment by ollieparanoid** [comment]
> Nexmon itself can be installed on the devices, correct?

Yes.

> If you don't ship the patched blobs (or even the blobs at all), and it's the user himself that "tinker for its own personal use" then I think you are as safe as you can possibly be.

That sounds nice, but:
* We would have a lot of additional effort to test the patches against all possible firmware versions that could be on the device (vs. testing the patch against one version that we would ship)
* There needs to be a good way to find out which version is already installed, the safest way would probably be hashing the firmware files. So we end up with hashes for all possible combinations of firmware versions and patches we provide, which doesn't scale well.
* When can't apply the patch for some reason, we can't implement a user-friendly way to overcome this (because flashing back the original firmware partition files is the only option, right?)

**Comment by mirh** [comment]
> We would have a lot of additional effort to test the patches against all possible firmware versions that could be on the device (vs. testing the patch against one version that we would ship)

Yes, you *can* ship that, and have this certainty
Or you *can* stick with what's on device to avoid whatever mumbo-scary-legal-jumbo you may fear. 

Then anyway, as I said, if that's really firmware (and not just blobs) I don't see what's the point of testing. 
I mean, there could be *bugs* with, say, older versions, sure.. But if they shipped that code once, I'd bet that basic functionality of whatever the component is always guaranteed (and then from there you could always proceed with #797 btw)

The point anyway would be that yes, you can ship what you want and 
Or you can use the one in the device, if you are wary 

> There needs to be a good way to find out which version is already installed, the safest way would probably be hashing the firmware files.

Speaking (again, duh) of wifi fw that I know, the version is patently [there](https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/tree/brcm/bcm4329-fullmac-4.bin) in the end. 
As for radio, I know that since the [dawn of time](https://android.googlesource.com/platform/hardware/ril/+/donut-release/libril/ril.cpp#2952) the version is always being neatly displayed (at least on android)...
So no need to "bruteforce it"

> When can't apply the patch for some reason, we can't implement a user-friendly way to overcome this (because flashing back the original firmware partition files is the only option, right?)

I dunno. I advanced the quibble that partitions *might* not be writable (since I know it's among "the legends" surrounding S-ON, even though the subject is so much confusing and contradictory for me) but even striving to come out with objections, to the best of my knowledge that only applies to radio for HTC devices*. 

I cannot instead see why or how /vendor would have to be treated differently than /system. 
...
Anyway, I just checked the vendor image from my XZ2c.. And holy moly isn't it a mess**. 
Long story short anyway (and what you could call my point): why can't a developer just choose whatever path for their bins to use?

*(and even putting aside S-*OFF* is possible in like almost every of them) even then.. it's not like there has ever been anybody else, other than the OEM, providing radio firmware?
The *only* case where this wouldn't hold, is with the tentative OsmocomBB project...... Which \*to be honest\*, at the moment, seems even more of a limited use case then even my "save few MBs" finickiness. 
**there is firmware, there are blobs, there are libraries, calibration files... No shit one wouldn't want to use it

**Comment by ollieparanoid** [comment]
> it's not like there has ever been anybody else, other than the OEM, providing radio firmware?
The *only* case where this wouldn't hold, is with the tentative OsmocomBB project...... Which \*to be honest\*, at the moment, seems even more of a limited use case then even my "save few MBs" finickiness. 

I think this sums up well where we have our different standpoint. And I get yours. Replacing the firmware stuff is future music, let's just use these few extra MBs today. I agree that this completely makes sense short term.

My standpoint is the long term vision. Security bugs in wifi and cellular modem **are real**, and they allow attackers to take over the whole device. We need to do something about that, and the best approach we have so far are binary patching and eventually replacing the firmware with open source code. You see, this is not some optional feature that power users would turn on somehwere hidden in the settings. It's an essential feature to make the phone reasonably safe to use. It should be the default for all users, especially the ones that don't know the technical details.

For that reason I believe we should not make an architecture decision that gives us problems in the future with modifying the firmware files. And there will be problems. There is so much that can go wrong when unconditionally applying binary patches to files without knowing the previous version. Who knows if the user had installed some experimental ROM before, that had modified firmware files for whichever reason (because if we can do it, why would other OSes not want to do it as well)? How do you solve incremental updates at all if you don't know the previous version? How do we even know if we already applied an update or not? When do we apply that update, at every boot? What if any of that goes wrong, we need a good solution that is not "try to find the vendor's firmware file and teach yourself how to flash it back with fastboot on the 
command line" and "well, if the init script is stuck at that step, just reinstall postmarketOS".

If we can only have the '"save few MBs" finickiness' as you call it by closing the door to this essential feature that we will need in the future, then it changes the project for the worse in my opinion.

> Long story short anyway (and what you could call my point): why can't a developer just choose whatever path for their bins to use?

Because good code and good systems need to be consistent! That's why we have standards like PEP8, FHS and POSIX. So what we do is put our heads together to come up with technical arguments that make good standardization. And when we have these, we apply them across whole projects, ideally with automated tests to make sure they stay consistent. Time passes by, and if people come up with better arguments in favor of changing the standards, then we do these changes and we do them *across the whole code base*.

**Comment by mirh** [comment]
> I think this sums up well where we have our different standpoint. And I get yours. Replacing the firmware stuff is future music, let's just use these few extra MBs today. I agree that this completely makes sense short term.

Ehrm.. Maybe the.. "logical steps" weren't explicit enough, but that's only true: 
\* some one single directive had really to be enforced universally
\* assuming there ***really*** existed such cases I hypothesized
\* there didn't exist devices with "private" (or if any, at most write only) fw

> (because if we can do it, why would other OSes not want to do it as well)? 

Because seriously, if an android developer cared for these things, he'd be developing for gnu/linux... like here. 
I don't think even replicant managed to do it, with radio firmware. 

> There is so much that can go wrong when unconditionally applying binary patches to files without knowing the previous version. How do you solve incremental updates at all if you don't know the previous version? How do we even know if we already applied an update or not? When do we apply that update, at every boot?

Why should you apply the update at every boot if it's already there?
And why cannot hashes be used to check the right fw version you want?
Last but not least, why requiring user to flash x, y or z stock rom as a very minimum requirement for functioning (for example, because its bugged bootloader allowed custom kernels) is kinda ok, while if we were to talk about radio (which would avoid those bla bla bla pesky legal wonderings) world is falling?

> What if any of that goes wrong, we need a good solution that is not "try to find the vendor's firmware file and teach yourself how to flash it back with fastboot on the command line" and "well, if the init script is stuck at that step, just reinstall postmarketOS".

I fail to see how this is any different than "flash stock rom" that might be recommended for just about any other possible problem with pmOS. 

> Because good code and good systems need to be consistent! That's why we have standards like PEP8, FHS and POSIX. 

Of course. 
I'm not sure though why [whatever any device stupid limitation of the day is] its workaround can still be called a day, while "almost totally in the air" ideas should warrant such strict policy. 

Standards are nice and all, but I'm not even sure how you'd go at radio (I'd like to underline again I am stating nothing about /vendor instead)? A loopback device pointing at an image of the partition? 

> and eventually replacing the firmware with open source code. You see, this is not some optional feature that power users would turn on somehwere hidden in the settings. It's an essential feature to make the phone reasonably safe to use. It should be the default for all users, especially the ones that don't know the technical details.

Not sure why being less noob would make the benefits less tangible, but even assuming everything was perfect, totally.. I cannot see why that couldn't still fit into the proper device partition ("so that every OS can benefit of it")?
And that would be *the standard* on such phone.
